hubspot3/tickets.py

```diff
@@ -78,3 +90,171 @@ def get_all(self, limit: int = -1, **options) -> list:
+        # https://community.hubspot.com/t5/APIs-Integrations/Time-difference-between-ticket-s-quot-get-log-of-changes-quot/m-p/297909#M27982
+        **options
+    ):
+
+        finished = False
+        query_limit = 100  # max according to the docs
+        limited = limit > 0
+        if limited and limit < query_limit:
+            query_limit = limit
+        total_tickets = 0
+        while not finished:
+            params = {"count": query_limit}
+            if vid_offset:
+                params["objectId"] = vid_offset
+            if time_offset:
+                params["timestamp"] = int(time_offset)
+            params['changeType'] = recency_type
+            batch = self._call(
+                "change-log/tickets",
+                method="GET",
+                params=params,
+                doseq=True,
+                **options
+            )
+
+            total_tickets += len(batch)
+            finished = (len(batch) == 0) or (limited and total_tickets >= limit)
+            if len(batch) > 0:
+                vid_offset = batch[-1]['objectId']
+                time_offset = batch[-1]['timestamp']
+                changes = self._parse_changes(batch, query_limit, max_time_diff_ms)
+                yield changes
+
+    def _parse_changes(self, batch, query_limit, max_time_diff_ms):
+        # First lets group changes by deal id
+        changes_by_id = [(ticket_change['objectId'], ticket_change) for ticket_change in batch]
+        changes_by_id = sorted(changes_by_id, key=itemgetter(0))
+        changes_by_id = dict((key, list(value)) for key, value in
+                             itertools.groupby(changes_by_id, key=itemgetter(0)))
+        # Loop through ids in batches of 100 (limit of the batch API)
+        for index in range(int(len(changes_by_id) / query_limit) + 1):
+            # Get a batch of 100 elements
+            changes_by_id_batch = dict(list(changes_by_id.items())[index * query_limit:(index + 1)
+                                                                   * query_limit])
+            properties = []
+            for ticket_id, changes in changes_by_id_batch.items():
+                for ticket_id, change in changes:
+                    if change['changeType'] != TicketsClient.Recency.DELETED:
+                        properties += change['changes']['changedProperties']
+            properties = set(properties)
+            # Get the information of each change
+            tickets_history = self.get_batch_with_history(list(changes_by_id_batch.keys()),
+                                                          extra_properties=list(properties))
+            for ticket_id, ticket_history in tickets_history.items():
+                ticket_id = int(ticket_id)
+                properties = ticket_history['properties']
+                # Match the information of changes with each change
+                changes = changes_by_id_batch[ticket_id]
+                for ticket_id, change in changes:
```

Comment:
Again, the duplicated `ticket_id`. If you can't change this, you can discard it using an underscore.